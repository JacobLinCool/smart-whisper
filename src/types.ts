export enum WhisperSamplingStrategy {
	WHISPER_SAMPLING_GREEDY,
	WHISPER_SAMPLING_BEAM_SEARCH,
}

export type TranscribeFormat = "simple" | "detail";

/**
 * See {@link https://github.com/ggerganov/whisper.cpp/blob/00b7a4be02ca82d53ac69dd2dd438c16e2af7658/whisper.h#L433C19-L433C19} for details.
 */
export interface TranscribeParams<
	Format extends TranscribeFormat = TranscribeFormat,
	TokenTimestamp extends boolean = false,
> {
	strategy: WhisperSamplingStrategy;
	n_threads: number;
	n_max_text_ctx: number;
	offset_ms: number;
	duration_ms: number;

	translate: boolean;
	no_context: boolean;
	no_timestamps: boolean;
	single_segment: boolean;
	print_special: boolean;
	print_progress: boolean;
	print_realtime: boolean;
	print_timestamps: boolean;

	token_timestamps: TokenTimestamp;
	thold_pt: number;
	thold_ptsum: number;
	max_len: number;
	split_on_word: boolean;
	max_tokens: number;

	speed_up: boolean;
	debug_mode: boolean;
	audio_ctx: number;

	tdrz_enable: boolean;

	initial_prompt: string;

	/**
	 * Language code, e.g. "en", "de", "fr", "es", "it", "nl", "pt", "ru", "tr", "uk", "pl", "sv", "cs", "zh", "ja", "ko"
	 */
	language: string;

	suppress_blank: boolean;
	suppress_non_speech_tokens: boolean;

	temperature: number;
	max_initial_ts: number;
	length_penalty: number;

	temperature_inc: number;
	entropy_thold: number;
	logprob_thold: number;
	no_speech_thold: number;

	best_of: number;

	beam_size: number;

	format: Format;
}

export interface TranscribeSimpleResult {
	from: number;
	to: number;
	text: string;
}

/**
 * Represents a detailed result of transcription.
 */
export interface TranscribeDetailedResult<TokenTimestamp extends boolean>
	extends TranscribeSimpleResult {
	/** The detected spoken language. */
	lang: string;
	/** The confidence level of the transcription, calculated by the average probability of the tokens. */
	confidence: number;
	/** The tokens generated during the transcription process. */
	tokens: {
		/** The text of the token, for CJK languages, due to the BPE encoding, the token text may not be readable. */
		text: string;
		/** The ID of the token. */
		id: number;
		/** The probability of the token. */
		p: number;
		/** The start timestamp of the token, in milliseconds. Only available when `token_timestamps` of {@link TranscribeParams} is `true`. */
		from: TokenTimestamp extends true ? number : undefined;
		/** The end timestamp of the token, in milliseconds. Only available when `token_timestamps` of {@link TranscribeParams} is `true`. */
		to: TokenTimestamp extends true ? number : undefined;
	}[];
}

export type TranscribeResult<
	Format extends TranscribeFormat = TranscribeFormat,
	TokenTimestamp extends boolean = boolean,
> = Format extends "simple" ? TranscribeSimpleResult : TranscribeDetailedResult<TokenTimestamp>;
