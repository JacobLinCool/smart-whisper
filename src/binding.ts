process.env.GGML_METAL_PATH_RESOURCES =
	process.env.GGML_METAL_PATH_RESOURCES || path.join(__dirname, "../whisper.cpp/ggml/src");

import path from "node:path";
import { TranscribeFormat, TranscribeParams, TranscribeResult } from "./types";
const module = require(path.join(__dirname, "../build/Release/smart-whisper"));

/**
 * A external handle to a model.
 */
export type Handle = {
	readonly "": unique symbol;
};

export namespace Binding {
	/**
	 * Load a model from a whisper weights file.
	 * @param file The path to the whisper weights file.
	 * @param gpu Whether to use the GPU or not.
	 * @param callback A callback that will be called with the handle to the model.
	 */
	export declare function load(
		file: string,
		gpu: boolean,
		callback: (handle: Handle) => void,
	): void;

	/**
	 * Release the memory of the model, it will be unusable after this.
	 * @param handle The handle to the model.
	 * @param callback A callback that will be called when the model is freed.
	 */
	export declare function free(handle: Handle, callback: () => void): void;

	/**
	 * Transcribe a PCM buffer.
	 * @param handle The handle to the model.
	 * @param pcm The PCM buffer.
	 * @param params The parameters to use for transcription.
	 * @param finish A callback that will be called when the transcription is finished.
	 * @param progress A callback that will be called when a new result is available.
	 */
	export declare function transcribe<
		Format extends TranscribeFormat,
		TokenTimestamp extends boolean,
	>(
		handle: Handle,
		pcm: Float32Array,
		params: Partial<TranscribeParams<Format, TokenTimestamp>>,
		finish: (results: TranscribeResult<Format, TokenTimestamp>[]) => void,
		progress: (result: TranscribeResult<Format, TokenTimestamp>) => void,
	): void;

	export declare class WhisperModel {
		private _ctx;
		constructor(handle: Handle);
		get handle(): Handle | null;
		get freed(): boolean;
		/**
		 * Release the memory of the model, it will be unusable after this.
		 * It's safe to call this multiple times, but it will only free the model once.
		 */
		free(): Promise<void>;
		/**
		 * Load a model from a whisper weights file.
		 * @param file The path to the whisper weights file.
		 * @param gpu Whether to use the GPU or not.
		 * @returns A promise that resolves to a {@link WhisperModel}.
		 */
		static load(file: string, gpu?: boolean): Promise<WhisperModel>;
	}
}

/**
 * The native binding for the underlying C++ addon.
 */
export const binding: typeof Binding = module;
