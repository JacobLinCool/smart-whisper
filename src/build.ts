import os from "node:os";
import { execSync } from "node:child_process";

type ComputeBackend = "cpu" | "accelerate" | "metal" | "clblast" | "openblas";

const cfg = config();

export const sources = cfg.sources.join(" ");
export const defines = cfg.defines.join(" ");
export const libraries = cfg.libraries.join(" ");

function config(): {
	sources: string[];
	defines: string[];
	libraries: string[];
} {
	if (process.env.BYOL) {
		return {
			sources: [],
			defines: [],
			libraries: [process.env.BYOL],
		};
	}

	const COMPUTE_BACKEND: ComputeBackend =
		(process.env.COMPUTE_BACKEND as ComputeBackend | undefined) ?? infer_backend();

	const cfg = {
		sources: [
			"whisper.cpp/src/whisper.cpp",
			"whisper.cpp/ggml/src/ggml.c",
			"whisper.cpp/ggml/src/ggml-alloc.c",
			"whisper.cpp/ggml/src/ggml-backend.c",
			"whisper.cpp/ggml/src/ggml-quants.c",
			"whisper.cpp/ggml/src/ggml-aarch64.c",
		] as string[],
		defines: [] as string[],
		libraries: [] as string[],
	};

	switch (COMPUTE_BACKEND) {
		case "accelerate": {
			cfg.defines.push("GGML_USE_ACCELERATE");

			cfg.libraries.push('"-framework Foundation"');
			cfg.libraries.push('"-framework Accelerate"');
			break;
		}
		case "metal": {
			cfg.sources.push("whisper.cpp/ggml/src/ggml-metal.m");

			cfg.defines.push("GGML_USE_ACCELERATE");
			cfg.defines.push("GGML_USE_METAL");

			cfg.libraries.push('"-framework Foundation"');
			cfg.libraries.push('"-framework Accelerate"');
			cfg.libraries.push('"-framework Metal"');
			cfg.libraries.push('"-framework MetalKit"');
			break;
		}
		case "openblas": {
			cfg.defines.push("GGML_USE_OPENBLAS");

			cfg.libraries.push("-lopenblas");
			break;
		}
		default: {
		}
	}

	return cfg;
}

function infer_backend(): ComputeBackend {
	let backend: ComputeBackend = "cpu";

	try {
		if (os.platform() === "darwin") {
			backend = "accelerate";
			if (os.arch() === "arm64") {
				backend = "metal";
			}
		} else if (os.platform() === "linux") {
			const has_libopenblas = !!execSync("ldconfig -p | grep libopenblas").toString().trim();
			if (has_libopenblas) {
				backend = "openblas";
			}
		}
	} catch {
		// if anything goes wrong, just use the default cpu backend
	}

	return backend;
}
