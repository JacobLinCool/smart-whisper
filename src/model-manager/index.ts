import path from "node:path";
import fs from "node:fs";
import os from "node:os";
import { Readable } from "node:stream";
import type { ReadableStream } from "node:stream/web";

const root = path.join(os.homedir(), ".smart-whisper");
const models = path.join(root, "models");
const ext = ".bin";

fs.mkdirSync(models, { recursive: true });

const BASE_MODELS_URL = "https://huggingface.co/ggerganov/whisper.cpp/resolve/main";

/**
 * MODELS is an object that contains the URLs of different ggml whisper models.
 * Each model is represented by a key-value pair, where the key is the model name
 * and the value is the URL of the model.
 */
export const MODELS = {
	tiny: `${BASE_MODELS_URL}/ggml-tiny.bin`,
	"tiny.en": `${BASE_MODELS_URL}/ggml-tiny.en.bin`,
	small: `${BASE_MODELS_URL}/ggml-small.bin`,
	"small.en": `${BASE_MODELS_URL}/ggml-small.en.bin`,
	base: `${BASE_MODELS_URL}/ggml-base.bin`,
	"base.en": `${BASE_MODELS_URL}/ggml-base.en.bin`,
	medium: `${BASE_MODELS_URL}/ggml-medium.bin`,
	"medium.en": `${BASE_MODELS_URL}/ggml-medium.en.bin`,
	"large-v1": `${BASE_MODELS_URL}/ggml-large-v1.bin`,
	"large-v2": `${BASE_MODELS_URL}/ggml-large-v2.bin`,
	"large-v3": `${BASE_MODELS_URL}/ggml-large-v3.bin`,
	"large-v3-turbo": `${BASE_MODELS_URL}/ggml-large-v3-turbo.bin`,
} as const;

export type ModelName = keyof typeof MODELS | (string & {});

/**
 * Downloads a ggml whisper model from a specified URL or shorthand.
 *
 * @param model - The model to download, specified either as a key of the {@link MODELS} object or as a URL.
 * @returns A promise that resolves to the name of the downloaded model.
 * @throws An error if the model URL or shorthand is invalid, or if the model fails to download.
 */
export async function download(model: ModelName): Promise<string> {
	let url = "",
		name = "";
	if (model in MODELS) {
		url = MODELS[model as keyof typeof MODELS];
		name = model;
	} else {
		try {
			url = new URL(model).href;
			name = new URL(url).pathname.split("/").pop() ?? "";
		} catch {}
	}

	if (!url) {
		throw new Error(`Invalid model URL or shorthand: ${model}`);
	}

	if (!name) {
		throw new Error(`Failed to parse model name: ${url}`);
	}

	if (check(name)) {
		return name;
	}

	const res = await fetch(url);
	if (!res.ok || !res.body) {
		throw new Error(`Failed to download model: ${res.statusText}`);
	}

	const stream = fs.createWriteStream(path.join(models, name.endsWith(ext) ? name : name + ext));
	Readable.fromWeb(res.body as ReadableStream<Uint8Array>).pipe(stream);

	return new Promise((resolve) => stream.on("finish", () => resolve(name)));
}

/**
 * Removes a locally downloaded model.
 * @param model - The name of the model to remove.
 */
export function remove(model: ModelName): void {
	if (check(model)) {
		fs.unlinkSync(path.join(models, model + ext));
	}
}

/**
 * Retrieves a list of model names that are available locally.
 * @returns An array of model names.
 */
export function list(): ModelName[] {
	const files = fs.readdirSync(models).filter((file) => file.endsWith(ext));
	return files.map((file) => file.slice(0, -ext.length));
}

/**
 * Checks if a model exists.
 * @param model - The name of the model.
 * @returns True if the model exists, false otherwise.
 */
export function check(model: ModelName): boolean {
	return fs.existsSync(path.join(models, model + ext));
}

/**
 * Resolves the absolute path of a model.
 * @param model - The name of the model.
 * @returns The resolved path of the model.
 * @throws Error if the model is not found.
 */
export function resolve(model: ModelName): string {
	if (check(model)) {
		return path.join(models, model + ext);
	} else {
		throw new Error(`Model not found: ${model}`);
	}
}

export const dir = { root, models };
