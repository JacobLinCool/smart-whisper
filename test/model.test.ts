import { manager, Whisper, WhisperModel } from "../src";
import { it, expect, beforeAll } from "vitest";
import fs from "node:fs";
import { decode } from "node-wav";

beforeAll(async () => {
	await manager.download("tiny");
});

it(
	"model should load",
	async () => {
		const fp = manager.resolve("tiny");
		const model = await WhisperModel.load(fp, false);
		expect(model.freed).toBe(false);
		expect(model.handle).not.toBe(null);
		await model.free();
		expect(model.freed).toBe(true);
		expect(model.handle).toBe(null);
	},
	{ timeout: 60_000 },
);

it(
	"model should work",
	async () => {
		const fp = manager.resolve("tiny");
		const whisper = new Whisper(fp, { gpu: false });
		const pcm = read_wav("./whisper.cpp/samples/jfk.wav");

		const { result } = await whisper.transcribe(pcm);
		const results = await result;
		expect(results[0].text).toMatchSnapshot();

		await whisper.free();
	},
	{ timeout: 60_000 },
);

function read_wav(file: string): Float32Array {
	const { sampleRate, channelData } = decode(fs.readFileSync(file));

	if (sampleRate !== 16000) {
		throw new Error(`Invalid sample rate: ${sampleRate}`);
	}
	if (channelData.length !== 1) {
		throw new Error(`Invalid channel count: ${channelData.length}`);
	}

	return channelData[0];
}
