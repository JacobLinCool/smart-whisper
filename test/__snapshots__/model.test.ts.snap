// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`model should work 1`] = `" And so, my fellow Americans, ask not what your country can do for you, ask what you can do for your country."`;
